storage/s3.go

```diff
@@ -382,6 +386,175 @@ func (s *S3) Get(
+		defer writer.Close()
+		// TODO: wire ctx up to eventStream.StreamCloser.Close
+		for event := range resp.EventStream.Reader.Events() {
+			switch e := event.(type) {
+			case *s3.RecordsEvent:
+				// TODO: how should we handle errors here?!
+				writer.Write(e.Payload)
+			// How the heck are these different?
+			case *s3.StatsEvent:
+			case *s3.ProgressEvent:
+			}
+		}
+	}()
+
+	decoder := json.NewDecoder(reader)
+	// decoder.DontCopyRawMessage()
+	for {
+		var record json.RawMessage
+		err := decoder.Decode(&record)
+		if err != nil {
+			if err != io.EOF {
+				resultch <- &SelectRecord{
+					Err: err,
+				}
+			}
+			return
+		}
+		resultch <- &SelectRecord{
+			Record: record,
+		}
+	}
+	err := resp.EventStream.Err()
+	if err != nil {
+		resultch <- &SelectRecord{
+			Err: err,
+		}
+	}
+}
+
+func (s *S3) selectWorker(ctx context.Context, input s3.SelectObjectContentInput, objch <-chan *Object, resultch chan<- *SelectRecord) {
+	for object := range objch {
+		// TODO: reintroduce
+		if object.Type.IsDir() { // || errorpkg.IsCancelation(object.Err) {
+			continue
+		}
+
+		if err := object.Err; err != nil {
+			resultch <- &SelectRecord{
+				Err: object.Err,
+			}
+			continue
+		}
+
+		if object.StorageClass.IsGlacier() {
+			// TODO: this isn't fatal - how to signal?
+			resultch <- &SelectRecord{
+				Err: fmt.Errorf("object '%v' is on Glacier storage", object),
+			}
+			continue
+		}
+
+		url := object.URL
+		input.Bucket = aws.String(url.Bucket)
+		input.Key = aws.String(url.Path)
+
+		out, err := s.api.SelectObjectContentWithContext(ctx, &input)
+		if err != nil {
+			resultch <- &SelectRecord{
+				Err: err,
+			}
+		} else {
+			s.selectStream(out, resultch)
+		}
+	}
+}
+
+type Query struct {
+	ExpressionType  string
+	Expression      string
+	CompressionType string
+}
+
+func writev(to io.Writer, vector ...[]byte) (err error) {
+	for _, vec := range vector {
+		_, err = to.Write(vec)
+		if err != nil {
+			break
+		}
+	}
+	return
+}
+
+func (s *S3) Select(ctx context.Context, objch <-chan *Object, to io.Writer, query *Query, concurrency int) error {
+	if s.dryRun {
+		return nil
+	}
+
+	baseInput := s3.SelectObjectContentInput{
+		ExpressionType: aws.String(query.ExpressionType),
+		Expression:     aws.String(query.Expression),
+		InputSerialization: &s3.InputSerialization{
+			CompressionType: aws.String(query.CompressionType),
+			JSON: &s3.JSONInput{
+				Type: aws.String("Lines"),
+			},
+		},
+		OutputSerialization: &s3.OutputSerialization{
+			JSON: &s3.JSONOutput{},
+		},
+		// RequestProgress: &s3.RequestProgress{
+		// 	// TODO: conditionally enable? for larger objects? ?
+		// 	Enabled: false,
+		// },
+	}
+
+	workerCtx, cancel := context.WithCancel(ctx)
+
+	resultch := make(chan *SelectRecord, concurrency)
+	var wg sync.WaitGroup
+	wg.Add(concurrency)
+	for worker := 0; worker < concurrency; worker++ {
+		go func() {
+			defer wg.Done()
+			s.selectWorker(workerCtx, baseInput, objch, resultch)
+		}()
+	}
```

Comment:
It seems like it'd be nice to use `parallel` for the workers here, but when I looked at it it looked like I couldn't stream tasks into it without buffering as many tasks as would exist. I'm not convinced my understanding is correct, though. I'm happy to rewrite to use `parallel` (I was just trying to get this working), but I'd like to figure out the appropriate concurrency model for the aforementioned design problem around wanting to select from many many files in the case that there are fewer matches (to try and vastly speed up needle-in-a-haystack selects).

Reply:
I sketched a prototype to demonstrate `parallel.Run` usage [here](https://github.com/peak/s5cmd/tree/select-proto1).

There may be some rough edges but overall, I think we can use the same structure as the `cp` command. Please take a look.

Reply:
Okay, pulled that in and refactored to correctly segment records.

Reply:
Thanks.